rfcs/038-matcher-versioning/README.md

```diff
@@ -0,0 +1,155 @@
+
+    ![](matcher_1.png)
+
+    The merger will take the maximum timestamp of each group to create the merged work, so it will create four works all with timestamp t=0.
+
+2.  Now suppose we realise these are paired incorrectly, and should actually be AD and BC.
+    We update A at time t=1, B at time t=2, C at time t=3, D at time t=4.
+    However, due to best-effort ordering, we don't process these updates in the correct order.
+
+3.  Suppose we process D, which was updated at time t=4.
+
+    ![](matcher_2.png)
+
+    All four works were processed together, so the merger creates four works all with timestamp t=4.
+
+4.  Then we process A and C, which were updated at time t=1 and t=3, respectively.
+
+    ![](matcher_3.png)
+
+    Again, the merger assigns the max timestamp t=4 to all four works.
+    This is equal to the previous max timestamp, so all four updates can be sent to the rest of the pipeline.
+
+5.  Finally, we try to process B, which was updated at t=2 – but we can't!
+
+    ![](matcher_4.png)
+
+    The matcher will send B and C together, so the merger uses the max timestamp t=3.
+    But there are already post-merger works for B/C from t=4, so these updates will be rejected.
+
+    It is impossible for the pipeline to process this update to B.
+    This update has become "stuck" because the post-merger works "remember" the works they used to be linked to.
+
+
+## Observations
+
+*   Works do sometimes become unlinked, and we need to handle it correctly.
+
+*   This problem gets worse when the pipeline is stable, and we don't reindex very often.
+    Every time we reindex, the pipeline "forgets" everything but the latest version of each source record, and any works with this issue get fixed.
+
+    This may explain some of the pipeline non-determinism issues we've seen – if one pipeline is "stuck", it would get different results from a fresh pipeline, regardless of how many times we retry or redrive.
+
+*   It's relatively easy to spot an issue of this with our current tooling – you see a warning message from the merger when it tries to index the relevant works, e.g.
+
+    ```
+    INFO  u.a.w.p.ElasticIndexer - Skipping h3xbcs2m because already indexed item has a higher version
+    INFO  u.a.w.p.ElasticIndexer - Skipping jz8d34qd because already indexed item has a higher version
+    INFO  u.a.w.p.ElasticIndexer - Skipping nymm52zc because already indexed item has a higher version
+    ```
+
+    But it can't be fixed once it's occurred, unless you manually edit the post-merger Elasticsearch indexes (scary).
+
+*   Until the matcher/merger, we have a 1:1 correspondence between "source record" and "catalogue Work".
+    After those services, we potentially break that link, and relying on it for ordering gets more difficult.
+
+
+## Proposed fix
+
+**We track the latest modified time of a *matcher graph*, and use that to version works post-matcher.**
```

Comment:
How (or where) do we do this? In the dynamo rows for graph nodes?

Reply:
Let me add a sentence on this.

Reply:
Ah, I see now - I think you could actually get this off the SQS message already https://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/API_ReceiveMessage.html

Reply:
Ooh, neat. I think I prefer it in our message – the SQS sending time could get messed up if we have to redrive messages.